### PR TITLE
210 docstrings

### PR DESCRIPTION
**Pull request by jarpri**

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import os
 import shutil
 
-import topomodelx
+#import topomodelx
```
Comment (ninamiolane) [inline docs/conf.py:6]: No left over comments. If it's not needed, remove it.

File: topomodelx/base/message_passing.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 
 
 class MessagePassing(torch.nn.Module):
-    """MessagePassing.
+    """Defines message passing.
```
Comment (ninamiolane) [inline topomodelx/base/message_passing.py:11]: Verb in infinitiv form, according to numpy-style guidelines for docstrings.

**Review by ninamiolane — APPROVED** [review]
Looking good, thanks for this important contribution!

I've added some minor comments.

I see in the docstrings that two conventions are used.

Sometimes the rank is denoted k, as in: n_k_cells and "k-cells".
Sometimes it is denoted r.

For consistency and for a smooth user experience, we should only use one.

Let's use r.

Thus rename: n_k_cells to n_r_cells and k-cells to r-cells.

Thanks!

**Comment by ninamiolane** [comment]
@jarpri there is a github conflict with the can_train notebook. Github conflicts are not fun to solve on github. I'd recommend removing the can_train.ipynb from this PR and doing another PR for it. Your call though!

**Comment by papamarkou** [comment]
Thank you @jarpri. @ninamiolane, is this ready to be merged?

File: topomodelx/nn/hypergraph/allset.py [hunk 3]
```diff
@@ -20,14 +20,14 @@ class AllSet(torch.nn.Module):
...
-    mlp_norm : bool, optional
-        Whether to apply input normalization in the MLP. Defaults to False.
+    n_layers : int, optional, default: 2
+        Number of AllSet layers in the network.
+    input_dropout : float, optional, default: 0.2
+        Dropout probability for the layer input.
+    mlp_num_layers : int, optional, default: 2
+        Number of layers in the MLP.
+    mlp_norm : bool, optional, default: False
+        Whether to apply input normalization in the MLP.
```
Comment (ffl096) [inline topomodelx/nn/hypergraph/allset.py:30]: Remove the `optional` here.

File: topomodelx/nn/cell/can_layer.py [hunk 4]
```diff
@@ -310,25 +316,29 @@ def reset_parameters(self) -> None:
...
-                = \phi^t(h_x^t, m_{x}^{(r)}), \forall x\in \mathcal C_r^{t+1}
-            \end{align*}
-
         Parameters
         ----------
-        x_0: torch.Tensor
-            Node signal of shape (num_nodes, in_channels_0).
-        neighborhood_0_to_0: torch.Tensor
-            Neighborhood matrix of shape (num_edges, 2).
+        x_0: torch.Tensor, shape = (num_nodes, in_channels_0)
```
Comment (ffl096) [inline topomodelx/nn/cell/can_layer.py:321]: Most of the time you use square brackets for the shape, here normal brackets. We should decide on one and be consistent with it.

File: topomodelx/nn/simplicial/sccn.py [hunk 5]
```diff
@@ -47,19 +49,12 @@ def forward(self, features, incidences, adjacencies):
 
         Parameters
         ----------
-        features: Dict[int, torch.Tensor],
-                length=max_rank+1,
-                shape=[n_rank_r_cells, channels]
+        features: Dict[int, torch.Tensor], length=max_rank+1, shape=[n_rank_r_cells, channels]
```
Comment (ffl096) [inline topomodelx/nn/simplicial/sccn.py:52]: `dict` in lowercase, please (same below) 🙂

File: topomodelx/nn/simplicial/sccn_layer.py [hunk 6]
```diff
@@ -149,25 +149,16 @@ def forward(self, features, incidences, adjacencies):
 
         Parameters
         ----------
-        features: Dict[int, torch.Tensor],
-                length=max_rank+1,
-                shape=[n_rank_r_cells, channels]
+        features: Dict[int, torch.Tensor], length=max_rank+1, shape=[n_rank_r_cells, channels]
```
Comment (ffl096) [inline topomodelx/nn/simplicial/sccn_layer.py:152]: `dict` in lowercase, please (same below) 🙂

File: topomodelx/nn/simplicial/sccnn.py [hunk 7]
```diff
@@ -197,23 +199,22 @@ def forward(self, x_all, laplacian_all, incidence_all):
...
+        incidence_all : tuple of tensors
+            Tuple of order 1 and 2 incidence matrices.
+            Shape of B1 = [n_nodes, n_edges].
+            Shape of B2 = [n_edges, n_faces].
 
         Returns
         -------
-        _ : tensor
-            shape = [n_nodes, 2]
+        _ : tensor, shape = [n_nodes, 2]
```
Comment (ffl096) [inline topomodelx/nn/simplicial/sccnn.py:217]:
```suggestion
        tensor, shape = [n_nodes, 2]
```

File: topomodelx/nn/simplicial/sccnn_layer.py [hunk 8]
```diff
@@ -207,17 +206,17 @@ def chebyshev_conv(self, conv_operator, conv_order, x):
...
+            Convolution operator e.g., the adjacency matrix, or the Hodge Laplacians.
+        conv_order : int
+            The order of the convolution.
+        x : torch.Tensor, shape = [n_simplices,num_channels]
+            Feature tensor.
+
+        Returns
         ------
-        x[:, :, k] = (conv_operator@....@conv_operator) @ x
+        X : torch.Tensor
```
Comment (ffl096) [inline topomodelx/nn/simplicial/sccnn_layer.py:218]: Single return values do not have a name.

```suggestion
        torch.Tensor
```

File: topomodelx/nn/simplicial/sccnn_layer.py [hunk 9]
```diff
@@ -249,31 +248,36 @@ def forward(self, x_all, laplacian_all, incidence_all):
...
+            - b1: torch.sparse, shape=[n_nodes,n_edges], node-to-edge incidence matrix,
+            - b2: torch.sparse, shape=[n_edges,n_triangles], edge-to-face incidence matrix.
+
         Returns
         -------
-        y: tuple (y_0,y_1,y_2)
-        - y_0: output on nodes
-        - y_1: output on edges
-        - y_2: output on triangles
+        y: tuple, shape = (y_0,y_1,y_2)
```
Comment (ffl096) [inline topomodelx/nn/simplicial/sccnn_layer.py:275]:
```suggestion
        tuple, shape = (y_0,y_1,y_2)
```

File: topomodelx/nn/simplicial/scconv_layer.py [hunk 10]
```diff
@@ -133,21 +133,26 @@ def forward(
...
+        adjacency_up_1_norm: torch.Tensor
+            Normalized upper adjacency matrix of rank 1.
+        adjacency_down_1_norm: torch.Tensor
+            Normalized down adjacency matrix of rank 1.
+        adjacency_down_2_norm: torch.Tensor
+            Normalized down adjacency matrix of rank 2.
+
+        Returns
+        -------
+        _ : tuple of tensors, shape = (x0_out, x1_out, x2_out)
```
Comment (ffl096) [inline topomodelx/nn/simplicial/scconv_layer.py:154]:
```suggestion
        tuple of tensors, shape = (x0_out, x1_out, x2_out)
```

File: topomodelx/nn/simplicial/scnn_layer.py [hunk 11]
```diff
@@ -136,17 +136,17 @@ def chebyshev_conv(self, conv_operator, conv_order, x):
...
+            Convolution operator e.g. adjacency matrix or the Hodge Laplacians.
+        conv_order : int
+            The order of the convolution
+        x : torch.Tensor, shape = [n_simplices,num_channels]
+            Input feature tensor.
 
         Return
         ------
-          x[:,:,k] = (conv_operator@....@conv_operator) @ x
+        X : torch.Tensor
```
Comment (ffl096) [inline topomodelx/nn/simplicial/scnn_layer.py:148]:
```suggestion
        torch.Tensor
```

**Review by ffl096 — CHANGES_REQUESTED** [review]
A few simple improvements, otherwise looks good to me!

File: topomodelx/nn/cell/can_layer.py (hunk 4 above)
Comment (papamarkou) [reply topomodelx/nn/cell/can_layer.py:321]: Just tagging @jarpri so that it noted.

File: topomodelx/nn/hypergraph/allset.py (hunk 3 above)
Comment (papamarkou) [reply topomodelx/nn/hypergraph/allset.py:30]: Tagging @jarpri.

File: topomodelx/nn/simplicial/sccn.py (hunk 5 above)
Comment (papamarkou) [reply topomodelx/nn/simplicial/sccn.py:52]: Tagging @jarpri.

**Comment by papamarkou** [comment]
@ffl096, if it would be possible to squash and merge, it would be the most efficient; rebasing the branch might be quite painful right now for @jarpri, due to not having repeated the process.

**Comment by papamarkou** [comment]
@ffl096, @ninamiolane, we communicated with @jarpri, who kindly offered to try in the next couple of days to do the rebasing (as a learning exercise). I recommended that Jaro makes a backup of his work, so that if anything goes wrong, he can simply resubmit a fresh PR. Let's give Jaro the chance to give it a go, and we may coordinate.

**Comment by ffl096** [comment]
Don‘t worry, nothing is ever lost. If everything goes wrong, we can always find the orphan commit of the current state of this pr and restore it :)

**Comment by jarpri** [comment]
@ffl096, I've completed the rebase process. Could you please review the changes? I don't understand these conflicts. Thanks for your help and patience all!

**Comment by papamarkou** [comment]
@ffl096, @ninamiolane  I can see @jarpri completed the changes you suggested and has done the rebase as well. Is this PR ok to merge, or are there any real pending conflicts?

**Review by ffl096 — APPROVED** [review]
Looks good to me! I am on holidays until next Wednesday, so if somebody else can resolve the conflicts and merge :)

**Comment by papamarkou** [comment]
> Looks good to me! I am on holidays until next Wednesday, so if somebody else can resolve the conflicts and merge :)

Thank you @ffl096, I took a look at the conflicts. If I understand well, this is just the result of prining some docstrings only in the the notebooks. @ninamiolane, is this considered an issue, do we need to do anything about it, or is it ok?

**Comment by papamarkou** [comment]
@ffl096, just a gentle follow up if this could be merged?

**Comment by ffl096** [comment]
This is ready to be merged now. No need to squash commits

**Comment by papamarkou** [comment]
> This is ready to be merged now. No need to squash commits

Thank you very much @ffl096!

**Comment by ffl096** [comment]
(Just for clarification: I can't merge anymore since I am now the last pusher to this pull request. Someone else has to hit the button)

**Comment by papamarkou** [comment]
Ok, I did it :) Thank you so much for your help @ffl096!

**Pull request merged**